include/albatross/src/stats/incomplete_gamma.hpp

```diff
@@ -0,0 +1,178 @@
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_STATS_INCOMPLETE_GAMMA_HPP_
+#define ALBATROSS_STATS_INCOMPLETE_GAMMA_HPP_
+
+/*
+ * The following methods use either Gaussian Quadrature or the power series
+ * expansion to integrate the incomplete gamma function
+ *
+ * https://en.wikipedia.org/wiki/Gaussian_quadrature
+ * https://en.wikipedia.org/wiki/Incomplete_gamma_function#Holomorphic_extension
+ *
+ * Many thanks to the GCEM package from which the majority of this was
+ * derived.
+ *
+ * https://github.com/kthohr/gcem/blob/master/include/gcem_incl/incomplete_gamma.hpp
+ */
+
+namespace albatross {
+
+namespace details {
+
+constexpr std::size_t INCOMPLETE_GAMMA_RECURSSION_LIMIT = 50;
+
+inline double incomplete_gamma_quadrature_inp_vals(double lb, double ub,
+                                                   std::size_t counter) {
+  assert(counter < gauss_legendre_50_points.size());
+  return (ub - lb) * 0.5 * gauss_legendre_50_points[counter] + 0.5 * (ub + lb);
+}
+
+inline double incomplete_gamma_quadrature_weight_vals(double lb, double ub,
+                                                      std::size_t counter) {
+  assert(counter < gauss_legendre_50_weights.size());
+  return (ub - lb) * 0.5 * gauss_legendre_50_weights[counter];
+}
+
+inline double incomplete_gamma_quadrature_fn(double x, double a,
+                                             double large_term) {
+  return exp(-x + (a - 1.) * log(x) - large_term);
+}
+
+inline double incomplete_gamma_quadrature_recursive(double lb, double ub,
+                                                    double a, double lg_term,
+                                                    std::size_t counter) {
+  if (counter < gauss_legendre_50_weights.size() - 1) {
+    return incomplete_gamma_quadrature_fn(
+               incomplete_gamma_quadrature_inp_vals(lb, ub, counter), a,
+               lg_term) *
+               incomplete_gamma_quadrature_weight_vals(lb, ub, counter) +
+           incomplete_gamma_quadrature_recursive(lb, ub, a, lg_term,
+                                                 counter + 1);
+  } else {
+    return incomplete_gamma_quadrature_fn(
+               incomplete_gamma_quadrature_inp_vals(lb, ub, counter), a,
+               lg_term) *
+           incomplete_gamma_quadrature_weight_vals(lb, ub, counter);
+  }
+}
+
+inline double incomplete_gamma_quadrature_lb(double a, double z) {
+  if (a > 1000) {
+    return std::max(0., std::min(z, a) - 11 * sqrt(a));
+  } else if (a > 800) {
+    return std::max(0., std::min(z, a) - 11 * sqrt(a));
```

Comment:
There seem to be several instances of conditional with the same execution.

Reply:
Yeah, kinda weird.  Not sure how those got selected, I cleaned it up a bit.